src/rust/engine/process_execution/src/lib.rs

```diff
@@ -341,13 +344,70 @@ pub struct ProcessMetadata {
 ///
 /// The result of running a process.
 ///
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Derivative, Clone, Debug, Eq)]
+#[derivative(PartialEq, Hash)]
 pub struct FallibleProcessResultWithPlatform {
   pub stdout_digest: Digest,
   pub stderr_digest: Digest,
   pub exit_code: i32,
-  pub platform: Platform,
   pub output_directory: hashing::Digest,
+  pub platform: Platform,
+  #[derivative(PartialEq = "ignore", Hash = "ignore")]
+  pub metadata: ProcessResultMetadata,
+}
+
+/// Metadata for a ProcessResult corresponding to the REAPI `ExecutedActionMetadata` proto. This
+/// conversion is lossy, but the interesting parts are preserved.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ProcessResultMetadata {
+  pub execution_time: Option<Duration>,
```

Comment:
Feedback appreciated: I'm not certain I chose the right set of timing metrics from https://github.com/bazelbuild/remote-apis/blob/9e72daff42c941baaf43a4c370e2607a984c58a7/build/bazel/remote/execution/v2/remote_execution.proto#L859-L895

I also considered `worker_start_timestamp` and `worker_completed_timestamp`, which captures all stages, not just process execution.

NB: I found that our local execution time under-reports what the dynamic UI shows:

```
11:43:23.81 [INFO] Starting: Resolving 3rdparty/python/constraints.txt
11:43:32.87 [INFO] Completed: Resolving 3rdparty/python/constraints.txt
11:43:32.87 [INFO] Some(Duration { secs: 7, nanos: 896468125 })
```

This is underreporting by 1.164 seconds, which I imagine is chroot materialization and cleanup? I don't know whether we should try to capture this overhead.

Reply:
> I also considered worker_start_timestamp and worker_completed_timestamp, which captures all stages, not just process execution.

Yea, we want all of the stages.

> This is underreporting by 1.164 seconds, which I imagine is chroot materialization and cleanup? I don't know whether we should try to capture this overhead.

Yes, we should. Would attempt to move the recording as far out as is relevant, which is effectively to the "top" of the relevant CommandRunner.

Reply:
Remote execution workers do capture input root setup time. That is the "input fetch" timestamps. https://github.com/pantsbuild/pants/blob/c9607601b482b3a4f99b266f419d8039572e59af/src/rust/engine/process_execution/bazel_protos/protos/bazelbuild_remote-apis/build/bazel/remote/execution/v2/remote_execution.proto#L841-L845.  Not sure how easy it would be to do that for the Pants local executor.